File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-python/src/parallel_consumer/_session.py
# Copyright (C) 2026 Antony Stubbs and contributors

"""The admin: one gRPC stream, one dispatch queue, and the threads that keep both moving.

Three threads and one rule between them. **The reader never blocks.** The stream carries the
control plane as well as records, so an admin that stops reading to slow the proxy down
head-of-line-blocks its own shutdown; the reader therefore only ever appends to the dispatch
queue and returns to the stream. Hand-out to the workers is a second thread's job, and reporting
outcomes back is a third's.

The dispatch queue's rules are the protocol's, not this library's invention - depth is the
proxy's own in-flight ceiling, hand-out is FIFO, overflow is a protocol violation rather than a
load condition, and on shutdown the queue is *released* rather than run or abandoned. They are
specified once for every language so that ten clients behave comparably.
"""

from __future__ import annotations

import dataclasses
import logging
import queue
import threading
import time
from typing import Any

import grpc

from . import _wire
from ._generated import proxy_pb2 as pb
from ._generated import proxy_pb2_grpc as pb_grpc
from ._pool import WorkerPool
from .errors import ProtocolViolation
from .options import ClientOptions
from .records import InboundRecord

log = logging.getLogger(__name__)

_HALF_CLOSE = None
"""Send-queue sentinel: the request iterator ends, which is how a client half-closes."""

_POLL = 0.05
"""Seconds a blocking wait parks for before re-checking the session's state."""

CAPABILITIES = (_wire.DISPATCH,)
"""What this client declares it implements. Each wave that adds a duty adds its token here."""


@dataclasses.dataclass(frozen=True)
class _Queued:
    """A dispatched record waiting for a free worker. The token stays opaque bytes throughout."""

    token: bytes
    record: InboundRecord


class Session:
    """One connection's worth of protocol: handshake, dispatch, report, drain.

    Constructing it performs the handshake, so the caller has an executor count to size the pool
    with the moment it returns. Nothing here opens a channel before it is constructed - and it is
    constructed after the worker pool exists, which is what keeps the fork safe.
    """

    def __init__(self, port: int, options: ClientOptions, pool: WorkerPool,
                 *, drain_timeout: float = 30.0) -> None:
        self._pool = pool
        self._drain_timeout = drain_timeout
        self._queue: queue.Queue[_Queued] = queue.Queue()
        self._sends: queue.Queue[Any] = queue.Queue()
        self._lock = threading.Lock()
        self._outstanding = 0
        self._draining = False
        self._stopping = False
        self._closed = False
        self._finished = threading.Event()
        self._failure: BaseException | None = None
        self._threads: list[threading.Thread] = []

        self._channel = grpc.insecure_channel(f"127.0.0.1:{port}")
        self._sends.put(pb.ClientMessage(
            configure=_wire.configure_message(options, CAPABILITIES)))
        # protoc's gRPC output carries no annotations, so the strict pass sees an untyped call here.
        # Suppressed at the site rather than by relaxing the module: this is the ONE crossing into
        # generated code, and pyproject's warn_unused_ignores deletes this line for us the day
        # protoc starts emitting stubs.
        self._call = pb_grpc.ProxyServiceStub(  # type: ignore[no-untyped-call]
            self._channel).Session(self._requests())

        configured = self._handshake()
        self.max_concurrency: int = configured.max_concurrency
        self.executor_count: int = configured.executor_count
        self.capabilities: frozenset[str] = frozenset(configured.capabilities)
        self._permits = threading.Semaphore(self.executor_count)

    # ---- lifecycle -----------------------------------------------------------------

    def start(self) -> None:
        """Starts the three threads. The pool's workers must already be running."""
        for name, target in (("pc-reader", self._read),
                             ("pc-handout", self._hand_out),
                             ("pc-reporter", self._report)):
            thread = threading.Thread(target=target, name=name, daemon=True)
            thread.start()
            self._threads.append(thread)

    def wait(self, timeout: float | None = None) -> bool:
        """Blocks until the session ends. Returns ``False`` if it is still running."""
        return self._finished.wait(timeout)

    def close(self) -> None:
        """Client-initiated shutdown: the same drain the proxy would ask for, unprompted.

        Stop handing records out, release everything queued, let executing records report, then
        half-close. The half-close *is* the shutdown signal - there is no request message for it.
        """
        if self._closed:
            return
        self._closed = True
        if not self._stopping:
            self._drain()
            self._stopping = True
            self._sends.put(_HALF_CLOSE)
        for thread in self._threads:
            thread.join(timeout=self._drain_timeout)
        self._channel.close()
        self._finished.set()

    @property
    def failure(self) -> BaseException | None:
        """The error that ended the session, if one did."""
        return self._failure

    # ---- the stream ----------------------------------------------------------------

    def _requests(self) -> Any:
        """The outbound half of the stream: everything the client says, in one place."""
        while True:
            message = self._sends.get()
            if message is _HALF_CLOSE:
                return
            yield message

    def _send(self, message: pb.ClientMessage) -> None:
        self._sends.put(message)

    def _handshake(self) -> pb.Configured:
        """Sends Configure, reads Configured. Inline, so the caller has the count on return.

        No deadline of its own: a proxy that accepts a connection and then says nothing is a
        liveness failure, and liveness is the lease's job rather than a one-off timer here.
        """
        try:
            first = next(self._call)
        except grpc.RpcError as refused:
            raise ProtocolViolation(
                f"the proxy refused the session: {refused.code().name} - {refused.details()}"
            ) from refused
        if first.WhichOneof("message") != "configured":
            raise ProtocolViolation(
                f"the handshake must answer Configure with Configured, got "
                f"{first.WhichOneof('message')}"
            )
        configured = first.configured
        for field in ("max_concurrency", "executor_count"):
            if not configured.HasField(field):
                raise ProtocolViolation(
                    f"Configured must always carry {field}; absence never means unlimited"
                )
        log.debug("configured: max_concurrency=%d executor_count=%d capabilities=%s",
                  configured.max_concurrency, configured.executor_count,
                  sorted(configured.capabilities))
        # Same crossing as the stub call above: the generated message class is untyped, so its
        # fields read as Any. The declared return type is the assertion; the ignore is the receipt.
        return configured  # type: ignore[no-any-return]

    def _read(self) -> None:
        """The reader. Appends and returns to the stream; it never waits for a worker."""
        try:
            for message in self._call:
                kind = message.WhichOneof("message")
                if kind == "dispatch":
                    self._on_dispatch(message.dispatch)
                elif kind == "shutdown":
                    self._on_shutdown()
                elif kind == "set_executor_count":
                    # Declared in the schema and never sent by a v1 proxy: the executor count is
                    # a pure function of connect-time configuration, computed once.
                    self._violated("the proxy sent SetExecutorCount, which a v1 proxy never sends")
                    return
                else:
                    self._violated(f"the proxy sent {kind}, which this session did not negotiate")
                    return
        except grpc.RpcError as broken:
            if not self._stopping:
                self._failure = broken
        finally:
            self._finished.set()

    def _on_dispatch(self, dispatch: pb.Dispatch) -> None:
        for dispatched in dispatch.records:
            with self._lock:
                self._outstanding += 1
                outstanding = self._outstanding
            if outstanding > self.max_concurrency:
                # Not a load condition: the queue's depth IS the proxy's declared in-flight
                # ceiling, so exceeding it means the proxy broke its own contract. Dropping the
                # record or growing the queue would hide that.
                self._violated(
                    f"the proxy dispatched {outstanding} records at once, over the "
                    f"max_concurrency of {self.max_concurrency} it declared"
                )
                return
            self._queue.put(_Queued(token=dispatched.token.SerializeToString(),
                                    record=_wire.inbound_record(dispatched)))

    def _on_shutdown(self) -> None:
        """Proxy-initiated drain: release the queue, let executing records finish, half-close."""
        log.debug("proxy asked for shutdown; draining")
        self._drain()
        self._stopping = True
        self._sends.put(_HALF_CLOSE)

    # ---- hand-out and reporting ----------------------------------------------------

    def _hand_out(self) -> None:
        """FIFO by arrival, and within one Dispatch by the order its records appear."""
        while not self._stopping:
            try:
                item = self._queue.get(timeout=_POLL)
            except queue.Empty:
                continue
            if self._draining or not self._await_permit():
                self._release(item)
                continue
            self._pool.submit(item.token, item.record)

    def _await_permit(self) -> bool:
        """Waits for a worker to come free, giving up if the session starts draining."""
        while not (self._draining or self._stopping):
            if self._permits.acquire(timeout=_POLL):
                return True
        return False

    def _report(self) -> None:
        """Sends each outcome as it arrives - independently, and out of dispatch order."""
        outcomes = self._pool.outcomes()
        while not self._stopping:
            try:
                token, outcome = outcomes.get(timeout=_POLL)
            except queue.Empty:
                continue
            except (OSError, ValueError):  # the pool's queue closed under us
                return
            self._send(pb.ClientMessage(report=_wire.report_message(token, outcome)))
            self._permits.release()
            self._settled()

    def _release(self, item: _Queued) -> None:
        """Returns a queued record the client never ran, attempt count unchanged."""
        if "shutdown" in self.capabilities:
            self._send(pb.ClientMessage(report=_wire.released_report(item.token)))
        # Without the shutdown capability negotiated there is no Released to send: the proxy
        # reclaims what it dispatched when the stream ends, and sending outside the negotiated
        # set would be this client's own protocol violation.
        self._settled()

    def _settled(self) -> None:
        with self._lock:
            self._outstanding -= 1

    def _drain(self) -> None:
        """Stops hand-out, releases the queue, waits for executing records to report."""
        self._draining = True
        while True:
            try:
                self._release(self._queue.get_nowait())
            except queue.Empty:
                break
        deadline = time.monotonic() + self._drain_timeout
        while time.monotonic() < deadline:
            with self._lock:
                if self._outstanding <= 0:
                    return
            time.sleep(_POLL)
        log.warning("drain timed out with %d record(s) still unreported", self._outstanding)

    def _violated(self, problem: str) -> None:
        """Fails the stream rather than absorbing a contract breach."""
        self._failure = ProtocolViolation(problem)
        log.error("protocol violation: %s", problem)
        self._stopping = True
        self._call.cancel()
        self._finished.set()

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-typescript/src/session.ts
// Copyright (C) 2026 Antony Stubbs and contributors

import { type ClientDuplexStream, credentials } from "@grpc/grpc-js";

import { ProtocolViolationError, SessionClosedError } from "./errors";
import { type Outcome, type RecordProcessor, applyProcessor } from "./outcome";
import { DispatchQueue, type QueuedRecord } from "./queue";
import type { InboundRecord } from "./records";
import type { ResolvedOptions } from "./options";
import {
  type ClientMessage,
  CommitMode as WireCommitMode,
  type Configure,
  type Configured,
  type DispatchRecord,
  ProcessingOrder as WireProcessingOrder,
  ProxyServiceClient,
  type ProxyMessage,
  type Token,
} from "./generated/parallelconsumer/proxy/v1/proxy";
import type { Duration } from "./generated/google/protobuf/duration";

/** The v1 capability tokens, as the specification names them. */
export const Capability = {
  dispatch: "dispatch",
  heartbeat: "heartbeat",
  manifest: "manifest",
  workerDeath: "worker-death",
  shutdown: "shutdown",
  terminal: "terminal",
  setExecutorCount: "set-executor-count",
} as const;

/**
 * WHAT THIS CLIENT DECLARES IT IMPLEMENTS - and it is deliberately not an empty list.
 *
 * An empty `capabilities` means the v1 BASELINE, which is a promise to implement all six baseline
 * tokens. A partial client declaring nothing would be granted heartbeat, manifest, worker-death and
 * shutdown duties it does not perform - and the first consequence is a lease that expires because
 * nobody is heartbeating, returning every in-flight record on a timer. So the declared set is a
 * constant naming exactly what is implemented, and the wave that implements a duty adds its token
 * here. Declaring by omission cannot happen, because omission is not what this list expresses.
 */
export const DECLARED_CAPABILITIES: readonly string[] = [Capability.dispatch];

/** The effective session, as the proxy reported it. Assert on what came back, never what was asked. */
export interface SessionInfo {
  /** The effective in-flight ceiling, and this client's dispatch-queue depth. */
  readonly maxConcurrency: number;
  /** How many executors to run. Sent once, never revised. */
  readonly executorCount: number;
  /** The negotiated intersection of this client's declared set and the proxy's own. */
  readonly capabilities: ReadonlySet<string>;
  readonly topics: readonly string[];
  readonly topicPattern: string | null;
}

/**
 * The admin: it holds the one gRPC stream, owns the dispatch queue, and runs the executors.
 *
 * CONCURRENCY MODEL - PROMISE CONCURRENCY ON ONE EVENT LOOP, NOT `worker_threads`, and the choice
 * is deliberate:
 *
 * - It keeps the user's function a CLOSURE. `worker_threads` starts a fresh isolate with no shared
 *   heap, so a function can only reach it as a module path or a source string - which would make
 *   this the one client in the fan-out whose processor cannot close over anything, contradicting
 *   the reference surface ("a closure/lambda/callable, never an importable name").
 * - It is what Node's own concurrency is FOR. `executorCount` concurrent async invocations give
 *   real parallelism for the I/O-bound work that is the overwhelming majority of Node workloads -
 *   an HTTP call, a database round trip - because the event loop overlaps exactly those.
 * - THE HONEST LIMIT, stated rather than hidden: a processor that does CPU work SYNCHRONOUSLY
 *   blocks the event loop, and therefore blocks this stream - reports stop flowing, and once
 *   heartbeats exist they stop too, which expires leases and redelivers records the worker is
 *   still holding. Node's answer to that is for the application to offload its own CPU work (its
 *   own worker pool, or a native async call) and `await` it here, which composes with this model;
 *   a library-imposed worker-thread executor would force EVERY user into module-path processors to
 *   serve the minority case. A worker-thread executor mode can be added later as an OPTION without
 *   touching the wire - `executor_count` is all the protocol knows.
 *
 * The transport itself never blocks on a processor: `onMessage` queues synchronously and returns,
 * and the executors are separate async loops. That is the KTD39 rule this model can guarantee, and
 * the paragraph above is the part it cannot.
 */
export class Session {
  private readonly queue: DispatchQueue;
  private readonly executors: Promise<void>[] = [];
  private readonly ended: Promise<void>;
  private endSession!: () => void;
  private failSession!: (error: unknown) => void;

  private polled = false;
  private closing = false;
  private finished = false;
  /** Set when this client cancels the call, so the resulting CANCELLED is not reported twice. */
  private failure: unknown = undefined;

  private constructor(
    private readonly options: ResolvedOptions,
    private readonly grpc: ProxyServiceClient,
    private readonly call: ClientDuplexStream<ClientMessage, ProxyMessage>,
    readonly info: SessionInfo,
  ) {
    this.queue = new DispatchQueue(info.maxConcurrency);
    this.ended = new Promise<void>((resolve, reject) => {
      this.endSession = resolve;
      this.failSession = reject;
    });
    // Marks the rejection handled the moment it exists. An application that never calls done() -
    // entirely reasonable - must not take an unhandled-rejection crash for a session error it has
    // no interest in; one that does call done() still receives it, because attaching a handler
    // does not consume the rejection.
    this.ended.catch(() => undefined);

    this.call.on("data", (message: ProxyMessage) => { this.onMessage(message); });
    this.call.on("error", (error: Error) => { this.fail(error); });
    this.call.on("end", () => { this.finish(); });
  }

  /**
   * Connects, sends `Configure`, and waits for `Configured` - the handshake, after which the
   * session is open. Only after `Configured` arrives may anything else happen on this stream.
   */
  static async open(options: ResolvedOptions, port: number): Promise<Session> {
    const grpc = new ProxyServiceClient(`127.0.0.1:${port}`, credentials.createInsecure());
    const call = grpc.session();
    try {
      const configured = await handshake(call, configureFrom(options));
      return new Session(options, grpc, call, sessionInfoFrom(configured));
    } catch (error) {
      call.cancel();
      grpc.close();
      throw error;
    }
  }

  /**
   * Starts consumption. NON-BLOCKING, and at most once per client.
   *
   * See `ParallelConsumerClient.poll` for why this returns rather than blocking, and for how the
   * session's end is observed instead.
   */
  poll(processor: RecordProcessor): void {
    if (this.polled) {
      throw new SessionClosedError("poll() may be called at most once per client");
    }
    if (this.finished) {
      throw new SessionClosedError("the session has already ended");
    }
    this.polled = true;
    for (let executor = 0; executor < this.info.executorCount; executor += 1) {
      this.executors.push(this.runExecutor(processor));
    }
  }

  /** Resolves when the session ends cleanly; rejects with what ended it otherwise. */
  done(): Promise<void> {
    return this.ended;
  }

  /**
   * Client-initiated shutdown: stop hand-out, let executing records finish and report, then
   * half-close the stream. The half-close IS the shutdown signal - there is no request message.
   *
   * WHAT HAPPENS TO QUEUED RECORDS, and why it is not what the guide's §5 says. That section has a
   * client report every queued record `Released`, but the `Released` outcome is gated behind the
   * `shutdown` capability, and this wave negotiates only `dispatch`. Sending it anyway would be the
   * ordinary un-negotiated-message violation. So: `Released` when `shutdown` is negotiated, and
   * otherwise the queue is discarded and the proxy reclaims those records through its
   * connection-loss path, with their attempt counts unchanged. Recorded as a specification defect
   * in `docs/inflight/clients/typescript.md`; the `Released` half arrives with the drain wave.
   */
  async close(): Promise<void> {
    if (this.closing) {
      await this.settled();
      return;
    }
    this.closing = true;

    const discarded = this.queue.close();
    if (discarded.length > 0) {
      this.options.onWarning(
        `discarding ${discarded.length} queued record(s) at close: the session negotiated ` +
          `[${[...this.info.capabilities].join(", ")}], so "${Capability.shutdown}" is off and ` +
          "Released may not be sent; the proxy returns them to scheduling",
      );
    }

    await Promise.allSettled(this.executors);
    if (!this.finished) {
      this.call.end();
    }
    await this.settled();
    this.grpc.close();
  }

  /** One executor: take a record, run the function, report the outcome, repeat. */
  private async runExecutor(processor: RecordProcessor): Promise<void> {
    for (;;) {
      const item = await this.queue.take();
      if (item === null) {
        return;
      }
      const outcome = await applyProcessor(processor, item.record);
      try {
        await this.send(reportFor(item, outcome));
      } catch (error) {
        // The stream is gone. The record is not this client's to resolve any more - the proxy's
        // reconnect window and lease machinery own it from here - so stop rather than spin.
        this.fail(error);
        return;
      } finally {
        this.queue.settle();
      }
    }
  }

  /** Everything the proxy says, and what this session is entitled to do about it. */
  private onMessage(message: ProxyMessage): void {
    if (message.dispatch !== undefined) {
      this.onDispatch(message.dispatch.records ?? []);
      return;
    }
    if (message.configured !== undefined) {
      // A second Configured is the proxy's truthful refusal of a reconfiguration. This client sends
      // one Configure and never reconfigures, so seeing one means the proxy is answering something
      // this client did not ask.
      this.options.onWarning("ignoring a second Configured: this client never reconfigures");
      return;
    }
    if (message.setExecutorCount !== undefined) {
      // Declared in the schema and never sent by a v1 proxy; the specification says a v1 client
      // treats one as a protocol violation rather than acting on it.
      this.fail(
        new ProtocolViolationError(
          "the proxy sent SetExecutorCount, which no v1 proxy sends and no v1 client implements",
        ),
      );
      return;
    }
    // Drop and Shutdown are gated by `manifest` and `shutdown`; neither is negotiated on this
    // session, so neither may arrive. The specification's receiver rule allows ignoring or failing;
    // ignoring keeps a proxy-side bug from destroying work this client is still doing.
    const kind = message.drop !== undefined ? "Drop" : message.shutdown !== undefined ? "Shutdown" : "an empty";
    this.options.onWarning(
      `ignoring ${kind} message: it is outside the negotiated capability set ` +
        `[${[...this.info.capabilities].join(", ")}]`,
    );
  }

  /** A wave: queue every record, in wave order. Never blocks, never drops. */
  private onDispatch(records: readonly DispatchRecord[]): void {
    for (const dispatched of records) {
      let queued: QueuedRecord;
      try {
        queued = toQueuedRecord(dispatched);
        this.queue.offer(queued);
      } catch (error) {
        // Queue overflow is the proxy exceeding its own declared ceiling. The specification says to
        // fail the stream with FAILED_PRECONDITION naming the count - which no gRPC CLIENT can do,
        // since only a server sets a status. Cancelling is the client-side equivalent, and the
        // count travels in the error this raises instead. See errors.ts.
        this.fail(error);
        return;
      }
    }
  }

  /** Serialized, backpressure-aware write. The callback fires once the frame is flushed. */
  private send(message: ClientMessage): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      this.call.write(message, (error?: Error | null) => {
        if (error) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  }

  private fail(error: unknown): void {
    if (this.finished) {
      return;
    }
    if (this.failure !== undefined) {
      return;
    }
    this.failure = error;
    this.finished = true;
    this.queue.close();
    this.call.cancel();
    this.failSession(error);
  }

  private finish(): void {
    if (this.finished) {
      return;
    }
    this.finished = true;
    this.queue.close();
    this.endSession();
  }

  /** Waits for the stream to end, bounded - a proxy that never completes must not hang close(). */
  private async settled(): Promise<void> {
    let timer: NodeJS.Timeout | undefined;
    const bound = new Promise<void>((resolve) => {
      timer = setTimeout(resolve, this.options.shutdownTimeoutMs);
      timer.unref();
    });
    await Promise.race([this.ended.catch(() => undefined), bound]);
    if (timer !== undefined) {
      clearTimeout(timer);
    }
  }
}

/** Sends `Configure` and waits for the `Configured` that opens the session. */
function handshake(
  call: ClientDuplexStream<ClientMessage, ProxyMessage>,
  configure: Configure,
): Promise<Configured> {
  return new Promise<Configured>((resolve, reject) => {
    const onData = (message: ProxyMessage) => {
      cleanup();
      if (message.configured === undefined) {
        reject(
          new ProtocolViolationError(
            "the proxy's first message on a fresh session was not Configured",
          ),
        );
        return;
      }
      resolve(message.configured);
    };
    const onError = (error: Error) => {
      cleanup();
      reject(error);
    };
    const onEnd = () => {
      cleanup();
      reject(new SessionClosedError("the proxy closed the stream before answering Configure"));
    };
    const cleanup = () => {
      call.off("data", onData);
      call.off("error", onError);
      call.off("end", onEnd);
    };
    call.on("data", onData);
    call.on("error", onError);
    call.on("end", onEnd);
    call.write({ configure });
  });
}

function configureFrom(options: ResolvedOptions): Configure {
  return {
    topics: options.topics === undefined ? undefined : [...options.topics],
    topicPattern: options.topicPattern,
    maxConcurrency: options.maxConcurrency,
    kafkaProperties: { ...options.kafkaProperties },
    capabilities: [...DECLARED_CAPABILITIES],
    ordering: wireOrdering(options.ordering),
    commitMode: wireCommitMode(options.commitMode),
    commitInterval: wireDuration(options.commitIntervalMs),
    defaultMessageRetryDelay: wireDuration(options.defaultMessageRetryDelayMs),
    pcInstanceTag: options.instanceTag,
  };
}

/**
 * Reads the effective configuration.
 *
 * `max_concurrency` and `executor_count` are always set despite their `optional` markers - a
 * `Configured` missing either is a protocol violation, and absence never means "unlimited". So they
 * are checked rather than defaulted: a fabricated ceiling here would turn a proxy bug into an
 * unbounded queue.
 */
function sessionInfoFrom(configured: Configured): SessionInfo {
  const maxConcurrency = configured.maxConcurrency;
  const executorCount = configured.executorCount;
  if (maxConcurrency === undefined || maxConcurrency < 1) {
    throw new ProtocolViolationError(
      "Configured carried no usable max_concurrency; absence never means unlimited",
    );
  }
  if (executorCount === undefined || executorCount < 1) {
    throw new ProtocolViolationError("Configured carried no usable executor_count");
  }
  return {
    maxConcurrency,
    executorCount,
    capabilities: new Set(configured.capabilities ?? []),
    topics: configured.topics ?? [],
    topicPattern: configured.topicPattern ?? null,
  };
}

/** The wire form of one dispatched record, with the token carried through untouched. */
function toQueuedRecord(dispatched: DispatchRecord): QueuedRecord {
  const token: Token | undefined = dispatched.token;
  const wire = dispatched.record;
  if (token === undefined || wire === undefined) {
    throw new ProtocolViolationError(
      "a dispatched record arrived without its token or its record",
    );
  }
  const record: InboundRecord = {
    topic: wire.topic ?? "",
    partition: wire.partition ?? 0,
    offset: wire.offset ?? 0n,
    key: wire.key ?? null,
    value: wire.value ?? null,
    attempt: dispatched.attempt ?? 0,
    lastFailureAt: dispatched.lastFailureAt ?? null,
    lastFailureReason: dispatched.lastFailureReason ?? null,
  };
  return { token, record };
}

/**
 * The report for one settled record. The token is the object that arrived, echoed byte-identically
 * because it is never taken apart. An outcome is always present: a `Report` without one is a
 * protocol violation, and this union has no arm that could produce it.
 */
function reportFor(item: QueuedRecord, outcome: Outcome): ClientMessage {
  if (outcome.kind === "success") {
    return {
      report: {
        token: item.token,
        success: {
          produce: outcome.produce.map((record) => ({
            topic: record.topic,
            key: record.key ?? undefined,
            value: record.value ?? undefined,
          })),
        },
      },
    };
  }
  return {
    report: {
      token: item.token,
      failure: { reason: outcome.reason ?? undefined },
    },
  };
}

function wireOrdering(ordering: ResolvedOptions["ordering"]): WireProcessingOrder | undefined {
  switch (ordering) {
    case "unordered":
      return WireProcessingOrder.PROCESSING_ORDER_UNORDERED;
    case "partition":
      return WireProcessingOrder.PROCESSING_ORDER_PARTITION;
    case "key":
      return WireProcessingOrder.PROCESSING_ORDER_KEY;
    default:
      return undefined;
  }
}

function wireCommitMode(mode: ResolvedOptions["commitMode"]): WireCommitMode | undefined {
  switch (mode) {
    case "periodic-consumer-sync":
      return WireCommitMode.COMMIT_MODE_PERIODIC_CONSUMER_SYNC;
    case "periodic-consumer-async":
      return WireCommitMode.COMMIT_MODE_PERIODIC_CONSUMER_ASYNCHRONOUS;
    default:
      return undefined;
  }
}

function wireDuration(milliseconds: number | undefined): Duration | undefined {
  if (milliseconds === undefined) {
    return undefined;
  }
  return {
    seconds: BigInt(Math.trunc(milliseconds / 1000)),
    nanos: (milliseconds % 1000) * 1_000_000,
  };
}
